include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h

```diff
@@ -228,71 +225,392 @@ struct __merge_sort_leaf_submitter<__internal::__optional_kernel_name<_LeafSortN
         // std::log2 may be prone to rounding errors on some architectures
         const std::int64_t __n_iter = sycl::ctz(__n_power2) - sycl::ctz(__leaf_size);
+
+        // Create container for storages with split-points on base diagonal
+        //  - each iteration should have their own container
+        __container_of_temp_storages_t __temp_sp_storages;
+
         for (std::int64_t __i = 0; __i < __n_iter; ++__i)
         {
-            __event_chain = __q.submit([&, __event_chain, __n_sorted, __data_in_temp](sycl::handler& __cgh) {
-                __cgh.depends_on(__event_chain);
-
-                oneapi::dpl::__ranges::__require_access(__cgh, __rng);
-                sycl::accessor __dst(__temp_buf, __cgh, sycl::read_write, sycl::no_init);
-
-                __cgh.parallel_for<_GlobalSortName...>(
-                    sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id) {
-                        const _IndexT __i_elem = __item_id.get_linear_id() * __chunk;
-                        const _IndexT __i_elem_local = __i_elem % (__n_sorted * 2);
-
-                        const _IndexT __offset = std::min<_IndexT>(__i_elem - __i_elem_local, __n);
-                        const _IndexT __n1 = std::min<_IndexT>(__offset + __n_sorted, __n) - __offset;
-                        const _IndexT __n2 = std::min<_IndexT>(__offset + __n1 + __n_sorted, __n) - (__offset + __n1);
-
-                        if (__data_in_temp)
-                        {
-                            const oneapi::dpl::__ranges::drop_view_simple __rng1(__dst, __offset);
-                            const oneapi::dpl::__ranges::drop_view_simple __rng2(__dst, __offset + __n1);
-
-                            const auto start = __find_start_point(__rng1, _IndexT{0}, __n1, __rng2, _IndexT{0}, __n2,
-                                                                  __i_elem_local, __comp);
-                            __serial_merge(__rng1, __rng2, __rng /*__rng3*/, start.first, start.second, __i_elem,
-                                           __chunk, __n1, __n2, __comp);
-                        }
-                        else
-                        {
-                            const oneapi::dpl::__ranges::drop_view_simple __rng1(__rng, __offset);
-                            const oneapi::dpl::__ranges::drop_view_simple __rng2(__rng, __offset + __n1);
-
-                            const auto start = __find_start_point(__rng1, _IndexT{0}, __n1, __rng2, _IndexT{0}, __n2,
-                                                                  __i_elem_local, __comp);
-                            __serial_merge(__rng1, __rng2, __dst /*__rng3*/, start.first, start.second, __i_elem,
-                                           __chunk, __n1, __n2, __comp);
-                        }
-                    });
-            });
+            if (2 * __n_sorted < __get_starting_size_limit_for_large_submitter<__value_type>())
```

Comment:
Does it make sense that the starting size limit is the same for "merge" as it is for segments of `merge_sort`?
I would expect that it would be a somewhat lower threshold for merge sort, since we can combine calculations of base diagonals across multiple segments of merge within merge sort in a single kernel.  

Perhaps it has more to do with the size between base segments than justifying the diagonal kernel, in which case it perhaps makes sense to have the same threshold. I think a discussion within a tech meeting would be helpful to consider this stuff.

Reply:
My experiments shows that it doesn't make sense.

Reply:
In offline discussion, I gave a hypothesis about this.  To put it briefly here... 
I think what makes sense with your experiments may be that when [sub]problems grow large enough to reach outside of some (last?) level of cache, global accesses make it slow enough to justify the large submitter.  This is a possible reasonable explanation for why the [sub]problem size is the key, and the same for both merge and merge_sort.

A possible next step could be to confirm this and more explicitly tie this threshold based on input type size and cache size.

Reply:
I'm OK with delaying investigation of this until after the release with a github issue, but it would be good not to lose this.  If the real performance data does not align with this understanding, then we would resolve it by improving our understanding of the threshold.

Reply:
TODO added